sips/cluster_consensus.md

```diff
@@ -0,0 +1,362 @@
+	// Processes a post-consensus message
+    ProcessPostConsensus(msg PartialSignatureMessages) 
+}
+
+type ClusterRunner struct {
+	Shares          map[ValidatorPubkey]Share
+    QBFTParams     QBFTParams
+	QBFTController *qbft.Controller
+	BeaconNetwork  *types.BeaconNetwork
+}
+
+// BeaconVote is the consensus data
+type BeaconVote struct {
+    BlockRoot         phase0.Root
+    Source            phase0.Checkpoint
+    Target            phase0.Checkpoint
+}
+
+// PartialSignatureMessage is a msg for partial Beacon chain related signatures (like partial attestation, block, randao sigs)
+type PartialSignatureMessage struct {
+	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
+	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
+	Signer           phase0.ValidatorIndex			// Changed from OperatorID
+}
+
+func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
+    bits := bitfield.New(duty.CommitteeLength)
+    bits.Set(duty.ValidatorCommitteeIndex)
+
+    return Attestation{
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        Source: vote.Source,
+        Target: vote.Target,
+        CommiteeIndex: duty.CommitteeIndex,
+        AggregationBits: bits,
+    }
+}
+
+func ConstructSyncCommittee(vote BeaconVote, duty AttesterDuty) SyncCommitteeMessage{
+    return SyncCommitteeMessage
+    {
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        ValidatorIndex: duty.ValidatorIndex,
+    }
+}
+```
+#### PartialSignatureMessages
+
+The current structure that we have in code can be unchanged.
+
+```go
+type PartialSignatureMessages struct {
+	Type     PartialSigMsgType
+	Slot     phase0.Slot
+	Messages []*PartialSignatureMessage
+}
+
+// PartialSignatureMessage is a msg for partial Beacon chain related signatures
+type PartialSignatureMessage struct {
+	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
+	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
+    Signer           OperatorID // TODO: there is an optimization where we don't have to duplicate this field
+}
+```
+
+We must have the following flow:
+1. When you `calculateExpectedRootsAndDomain` (similar to `sync_committee_aggregator`). Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
+2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`
+3. When processing the messages find all the roots that have quorums for a certain validator, mark them, and sumbit corresponding beacon data to beacon chain.
+4. For the next message received attempt to complete quorum for other roots.
+
+
+#### Happy Flow
+
+1. `Cluster` receives duties that match a certain slot. If the slot is higher then `highestDecidedSlot` starts consensus for the relevant roles, and initializes a `ClusterRunner` for the relevant Validators.
```

Comment:
@GalRogozinski I think we dont use highestDecidedSlot anymore, we should start consensus and then not broadcasting partial sig for a validator in case the slot is not higher than HighestAttestingSlot, correct?